adapters/repos/db/queue/scheduler.go

```diff
@@ -0,0 +1,490 @@
+	"os"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/weaviate/weaviate/adapters/repos/db/vector/common"
+	enterrors "github.com/weaviate/weaviate/entities/errors"
+)
+
+type Scheduler struct {
+	SchedulerOptions
+
+	queues struct {
+		sync.Mutex
+
+		m map[string]*queueState
+	}
+
+	// context used to close pending tasks
+	ctx      context.Context
+	cancelFn context.CancelFunc
+
+	activeTasks *common.SharedGauge
+
+	wg        sync.WaitGroup
+	chans     []chan Batch
+	triggerCh chan chan struct{}
+}
+
+type SchedulerOptions struct {
+	Logger logrus.FieldLogger
+	// Number of workers to process tasks. Defaults to the number of CPUs - 1.
+	Workers int
+	// The interval at which the scheduler checks the queues for tasks. Defaults to 1 second.
+	ScheduleInterval time.Duration
+	// Function to be called when the scheduler is closed
+	OnClose func()
+}
+
+func NewScheduler(opts SchedulerOptions) *Scheduler {
+	if opts.Logger == nil {
+		opts.Logger = logrus.New()
+	}
+	opts.Logger = opts.Logger.WithField("component", "queue-scheduler")
+
+	if opts.Workers <= 0 {
+		opts.Workers = max(1, runtime.GOMAXPROCS(0)-1)
+	}
+
+	if opts.ScheduleInterval == 0 {
+		it, err := time.ParseDuration(os.Getenv("QUEUE_SCHEDULER_INTERVAL"))
+		if err != nil {
+			it = 1 * time.Second
+		}
+		opts.ScheduleInterval = it
+	}
+
+	s := Scheduler{
+		SchedulerOptions: opts,
+		activeTasks:      common.NewSharedGauge(),
+	}
+	s.queues.m = make(map[string]*queueState)
+	s.triggerCh = make(chan chan struct{})
+
+	return &s
+}
+
+func (s *Scheduler) RegisterQueue(q Queue) {
+	s.queues.Lock()
+	defer s.queues.Unlock()
+
+	s.queues.m[q.ID()] = newQueueState(q)
+
+	q.Metrics().Registered(q.ID())
+}
+
+func (s *Scheduler) UnregisterQueue(id string) {
+	s.PauseQueue(id)
+
+	q := s.getQueue(id)
+	if q == nil {
+		return
+	}
+
+	// wait for the workers to finish processing the queue's tasks
+	q.activeTasks.Wait()
+
+	// the queue is paused, so it's safe to remove it
+	s.queues.Lock()
+	delete(s.queues.m, id)
+	s.queues.Unlock()
+
+	q.q.Metrics().Unregistered(q.q.ID())
+}
+
+func (s *Scheduler) Start() {
+	if s.ctx != nil {
+		// scheduler already started
+		return
+	}
+
+	s.ctx, s.cancelFn = context.WithCancel(context.Background())
+
+	// run workers
+	chans := make([]chan Batch, s.Workers)
+
+	for i := 0; i < s.Workers; i++ {
+		worker, ch := NewWorker(s.Logger, 5*time.Second)
+		chans[i] = ch
+
+		s.wg.Add(1)
+		f := func() {
+			defer s.wg.Done()
+
+			worker.Run(s.ctx)
+		}
+		enterrors.GoWrapper(f, s.Logger)
+	}
+
+	s.chans = chans
+
+	// run scheduler goroutine
+	s.wg.Add(1)
+	f := func() {
+		defer s.wg.Done()
+
+		s.runScheduler()
+	}
+	enterrors.GoWrapper(f, s.Logger)
+}
+
+func (s *Scheduler) Close() error {
+	if s == nil || s.ctx == nil {
+		// scheduler not initialized. No op.
+		return nil
+	}
+
+	// check if the scheduler is already closed
+	if s.ctx.Err() != nil {
+		return nil
+	}
+
+	// stop scheduling
+	s.cancelFn()
```

Comment:
just to validate, cancelling the context won't prevent all the active tasks to be completed or if it does it's not a problem to not complete all of them...

Reply:
Canceling the context will cancel all active tasks, as the context is passed to the workers.
This is not a problem because the tasks are supposed to be idempotent